File: frontend/src/constants/Application.ts
import { Dayjs } from "dayjs";
import { Customer } from "./Customer";
import { ArchObject } from "./ArchObject";

export type Application = {
  id: number;
  registration_date: Dayjs;
  customer_id: number;
  status_id: number;
  workflow_id: number;
  service_id: number;
  deadline: Dayjs;
  arch_object_id: number;
  arch_process_id?: number;
  updated_at: Dayjs;
  customer: Customer;
  archObjects: ArchObject[];
  number?: string;
  work_description?: string;
  service_name?: string;
  status_name?: string;
  status_code?: string;
  object_tag_id: number;
  object_tag_name?: number;
  arch_object_district?: string;
  incoming_numbers?: string;
  outgoing_numbers?: string;
  dp_outgoing_number?: string;
  tech_decision_id?: number;
};


export type FilterApplication = {
  pageSize: number;
  pageNumber: number;
  sort_by: string;
  sort_type: string;
  pin: string;
  customerName: string;
  date_start: string;
  date_end: string;
  service_ids: number[];
  status_ids: number[];
  address: string;
  number: string;
  district_id: number;
  deadline_day: number;
  tag_id: number;
  isExpired: boolean;
  isMyOrgApplication: boolean;
  withoutAssignedEmployee: boolean;
  employee_id: number;
  useCommon: boolean;
  common_filter: string;
  structure_ids: number[];
  incoming_numbers: string;
  outgoing_numbers: string;
  employee_arch_id?: number;
  dashboard_date_start?: string;
  dashboard_date_end?: string;
  issued_employee_id?: number;
  only_count: boolean;
  is_paid?: boolean;
  dp_outgoing_number?: string;
};

File: backend/Domain/Entities/Application.cs

namespace Domain.Entities
{
    public class Application : BaseLogDomain, IBaseDomain
    {
        public int Id { get; set; }
        public DateTime? RegistrationDate { get; set; }
        public int CustomerId { get; set; }
        public string CustomerName { get; set; }
        public string CustomerPin { get; set; }
        public int StatusId { get; set; }
        public int WorkflowId { get; set; }
        public int ServiceId { get; set; }
        public int? WorkflowTaskStructureId { get; set; }
        public string ServiceName { get; set; }
        public string? CreatedByName { get; set; }
        public string? AssignedEmployeesNames { get; set; }
        public string? UpdatedByName { get; set; }
        public DateTime? Deadline { get; set; }
        public int? ArchObjectId { get; set; }
        public string? ArchObjectName { get; set; }
        public string? ArchObjectAddress { get; set; }
        public string? ArchObjectDistrict { get; set; }
        public int? DistrictId { get; set; }
        public bool? IsPaid { get; set; }
        public string? Number { get; set; }
        public string StatusName { get; set; }
        public string StatusCode { get; set; }
        public string? StatusColor { get; set; }
        public string ObjectTagName { get; set; }
        public bool? CustomerIsOrganization { get; set; }
        public string CustomerAddress { get; set; }
        public string CustomerOrganizationTypeName { get; set; }
        public string CustomerOkpo { get; set; }
        public string CustomerDirector { get; set; }
        public int? MariaDbStatementId { get; set; }
        public string? WorkDescription { get; set; }
        public string? CustomerContacts { get; set; }
        public List<ArchObject> ArchObjects { get; set; }
        public string? AssigneeEmployees { get; set; }
        public List<string> CustomersInfo { get; set; }
        public int RegistryId { get; set; }
        public int? ObjectTagId { get; set; }
        public string RegistryName { get; set; }
        public int? ArchProcessId { get; set; }
        public int? TechDecisionId { get; set; }
        public DateTime? TechDecisionDate { get; set; }
        public string IncomingNumbers { get; set; }
        public string OutgoingNumbers { get; set; }
        public decimal SumWithoutDiscount { get; set; }
        public decimal TotalSum { get; set; }
        public decimal DiscountPercentage { get; set; }
        public decimal DiscountValue { get; set; }
        public decimal VatValue { get; set; }
        public decimal ServiceTaxValue { get; set; }
        public decimal VatPercentage { get; set; }
        public decimal ServiceTaxPercentage { get; set; }
        public bool HasDiscount { get; set; }
        public int CalculationUpdatedBy { get; set; }
        public int CalculationCreatedBy { get; set; }        
        public DateTime? CalculationCreatedAt { get; set; }
        public DateTime? CalculationUpdatedAt { get; set; }
        public decimal TotalPaid { get; set; }
        public int? DayCount { get; set; }
        public string DecisionPaperOutgoingNumber { get; set; }
        public string? ApplicationCode { get; set; }
        public bool? MatchedContact { get; set; }
        public Validation Validate()
        {
            var errors = new List<FieldError>();
            if (ServiceId == 0)
            {
                errors.Add(new FieldError { ErrorCode = nameof(ErrorCode.COMMON_ID_NOT_NULL), FieldName = nameof(ServiceId) });
            }
            if (errors.Count > 0)
            {
                return Validation.NotValid(errors);
            }
            return Validation.Valid();
        }
    }
}
